File: bench/microbench3.pl
#!/usr/bin/perl
use strict;
use warnings;

package Dim;
sub new {
	my $class = shift;
	return bless {x => shift, y => shift, z => shift}, $class;
}
sub f {
}

package main;

sub testSimpleLoop () {
	my $i = 0;
	for($i=0;$i < 10000000; $i++) {
	}
	return $i;
}


sub testLocalVariable() {
	my $y = 0;
	for (my $i=0; $i < 10000000; $i++) {
		$y = $i;
	}
}

my $global_x;
my $global_y;
my $global_z;

sub testGlobalVariable {
	for (my $i=0; $i < 10000000; $i++) {
		$global_y = $i;
	}
	return $global_y;
}

sub testStringAssignment {
	my $s = "";
	for (my $i = 0; $i < 10000000; $i++) {
		$s = "A";
	}
	return $s;
}

sub testIntegerOperation() {
	my $y = 0;
	for (my $i = 0; $i < 10000000; $i++) {
		$y += $i;
	}
	return $y;
}

sub testFloatOperation {
	my $f = 0.0;
	for (my $i = 0; $i < 10000000; $i++) {
		$f += 0.1;
	}
	return $f;
}


sub func0 {
}

sub testFunctionCall {
	my $i;
	for ($i = 0; $i < 10000000; $i++) {
		func0();
	}
	return $i;
}

sub testFunctionReturn {
	my $res = 0;
	for (my $i = 0; $i < 10000000; $i++) {
		$res = func1();
	}
	return $res;
}

sub func1 {
	return 1;
}

sub testMathFabs() {
	my $res = 0.0;
	my $i = 0;
	for ($i = 0; $i < 10000000; $i++) {
		$res = abs(-1.0);
	}
	return $i;
}

sub testCallFunctionObject{

}

sub testObjectCreation {
	my $i = 0;
	for ($i=0; $i < 10000000; $i++) {
		my $d = new Dim();
	}
	return $i;
}

sub testFieldVariable {
	my $d = new Dim();
	my $i = 0;
	for ($i=0; $i < 10000000; $i++) {
		($d)->{y} = 1;
	}
	return $d;
}


sub testMethodCall {
	my $d = new Dim();
	my $i = 0;
	for ($i=0; $i < 10000000; $i++) {
		($d)->f();
	}
	return $d;
}

sub fibo {
	my ($n) = @_;
	if ($n < 3) {
		return 1;
	}
	return fibo($n - 1) + fibo($n - 2);
}

sub mops {
	my ($score) = @_;
	if ($score < 1) {
		return 0;
	}
	#return (10000000 / ($score)) / 100.0;
	return (10000 / ($score)) / 1000;
}

my ($score, $period, $delta);
my ($user1, $system, $cursor, $csystem);
my ($user2, $system2, $cursor2, $csystem2);
($user1, $system, $cursor, $csystem) = times;
testLocalVariable();
($user2, $system2, $cursor2, $csystem2) = times;
$period = $user2 - $user1;
$score = mops($period);
print "SimpleLoop:$score ", $user2-$user1, "[sec]\n";
$delta = $period;

($user1, $system, $cursor, $csystem) = times;
testLocalVariable();
($user2, $system2, $cursor2, $csystem2) = times;
$score = mops($user2-$user1 - $delta);
print "LocalVariable:$score ", $user2-$user1, "[sec]\n";


($user1, $system, $cursor, $csystem) = times;
testGlobalVariable();
($user2, $system2, $cursor2, $csystem2) = times;
$score = mops($user2-$user1 - $delta);
print "GlobalVariable:$score ", $user2-$user1, "[sec]\n";

($user1, $system, $cursor, $csystem) = times;
testStringAssignment();
($user2, $system2, $cursor2, $csystem2) = times;
$score = mops($user2-$user1 - $delta);
print "StringAssignment:$score ", $user2-$user1, "[sec]\n";

($user1, $system, $cursor, $csystem) = times;
testIntegerOperation();
($user2, $system2, $cursor2, $csystem2) = times;
$score = mops($user2-$user1 - $delta);
print "IntegerOperation:$score ", $user2-$user1, "[sec]\n";

($user1, $system, $cursor, $csystem) = times;
testFloatOperation();
($user2, $system2, $cursor2, $csystem2) = times;
$score = mops($user2-$user1 - $delta);
print "FloatOperation:$score ", $user2-$user1, "[sec]\n";

($user1, $system, $cursor, $csystem) = times;
testFunctionCall();
($user2, $system2, $cursor2, $csystem2) = times;
$score = mops($user2-$user1 - $delta);
print "FunctionCall:$score ", $user2-$user1, "[sec]\n";

($user1, $system, $cursor, $csystem) = times;
testFunctionReturn();
($user2, $system2, $cursor2, $csystem2) = times;

print "FunctionReturn:$score ", $user2-$user1, "[sec]\n";

($user1, $system, $cursor, $csystem) = times;
testMathFabs();
($user2, $system2, $cursor2, $csystem2) = times;
$score = mops($user2-$user1 - $delta);
print "NativeFunction: (none)\n";

($user1, $system, $cursor, $csystem) = times;
testObjectCreation();
($user2, $system2, $cursor2, $csystem2) = times;
$score = mops($user2-$user1 - $delta);
print "ObjectCreation:$score ", $user2-$user1, "[sec]\n";

($user1, $system, $cursor, $csystem) = times;
testFieldVariable();
($user2, $system2, $cursor2, $csystem2) = times;
$score = mops($user2-$user1 - $delta);
print "FieldVariable:$score ", $user2-$user1, "[sec]\n";

($user1, $system, $cursor, $csystem) = times;
testMethodCall();
($user2, $system2, $cursor2, $csystem2) = times;
$score = mops($user2-$user1 - $delta);
print "MethodCall:$score ", $user2-$user1, "[sec]\n";

($user1, $system, $cursor, $csystem) = times;
fibo(36);
($user2, $system2, $cursor2, $csystem2) = times;
$score = ($user2-$user1);
print "fib(36):$score ", $user2-$user1, "[sec]\n";

File: bench/microbench3.py
import time, math

def testSimpleLoop():
	i = 0
	while i < 10000000:
		i += 1
	return i

def testLocalVariable():
	i = 0
	y = 0
	while i < 10000000:
		y = i
		i += 1
	return y

global_x = 0
global_y = 0
global_z = 0

def testGlobalVariable():
	global global_y
	i = 0
	while i < 10000000:
		global_y = 1
		i += 1
	return global_y

def testStringAssignment():
	i = 0
	y = ""
	while i < 10000000:
		y = "2"
		i += 1
	return y

class Dim:
	x = 0
	y = 0
	z = 0
	def __init__(self):
		pass

	def f(self):
		pass

def testObjectCreation():
	i = 0
	while i < 10000000:
		d = Dim()
		i += 1
	return i

def testFieldVariable():
	d = Dim()
	i = 0
	while i < 10000000:
		d.y = i
		i += 1
	return d

def testMethodCall():
	d = Dim()
	i = 0
	while i < 10000000:
		d.f()
		i += 1
	return d

def testIntegerOperation():
	i = 0
	y = 0
	while i < 10000000:
		y += 1
		i += 1
	return y

def testFloatOperation():
	i = 0
	y = 0.0
	while i < 10000000:
		y += 0.2
		i += 1
	return y

def func0():
	pass

def testFunctionCall():
	i = 0
	while i < 10000000:
		func0()
		i += 1
	return i

def func1():
	return 1

def testFunctionReturn():
	i = 0
	y = 0
	while i < 10000000:
		y = func1()
		i += 1
	return y

def testMathFabs():
	i = 0
	f = 1.0
	while i < 10000000:
		i += 1
		f = math.fabs(-1.0)
	return f

def testCallFunctionObject():
	i = 0
	f = func0
	while i < 10000000:
		i += 1
		f()
	return i

def testFibo(n):
	if n < 3 : return 1
	return testFibo(n-1) + testFibo(n-2)

def mops(score):
	return 10 * 1000 / score

def run():
	s = time.time()
	testSimpleLoop()
	period = (time.time() - s) * 1000
	print "SimpleLoop[mpos,ms]: ", mops(period), period
	delta = period
	##
	s = time.time()
	testLocalVariable()
	period = ((time.time() - s) * 1000)
	score = period - delta
	print "LocalVariable[mpos,ms]:", mops(score), period
	##
	s = time.time()
	testGlobalVariable()
	period = ((time.time() - s) * 1000)
	score = period - delta
	print "GlobalVariable[mpos,ms]:", mops(score), period
	##
	s = time.time()
	testStringAssignment()
	period = ((time.time() - s) * 1000)
	score = period - delta
	print "StringAssignment[mpos,ms]:", mops(score), period
	##
	s = time.time()
	testIntegerOperation()
	period = ((time.time() - s) * 1000)
	score = period - delta
	print "IntegerOperation[mpos,ms]:", mops(score), period
	##
	s = time.time()
	testFloatOperation()
	period = ((time.time() - s) * 1000)
	score = period - delta
	print "FloatOperation[mpos,ms]:", mops(score), period
	##
	s = time.time()
	testFunctionCall()
	period = ((time.time() - s) * 1000)
	score = period - delta
	print "FunctionCall[mpos,ms]:", mops(score), period
	##
	s = time.time()
	testFunctionReturn()
	period = ((time.time() - s) * 1000)
	score = period - delta
	print "FunctionReturn[mpos,ms]:", mops(score), period
	##
	s = time.time()
	testMathFabs()
	period = ((time.time() - s) * 1000)
	score = period - (delta)
	print "NaitiveFunction[mpos,ms]:", mops(score), period
	##
	s = time.time()
	testCallFunctionObject()
	period = ((time.time() - s) * 1000)
	score = period - (delta)
	print "CallFunctionObject[mpos,ms]:", mops(score), period
	##
	print "Object-Oriented Programming"
	s = time.time()
	testObjectCreation()
	period = (time.time() - s) * 1000
	score = period - delta
	print "ObjectCreation[mpos,ms]:", mops(score), period
	s = time.time()
	testFieldVariable()
	period = (time.time() - s) * 1000
	score = period - delta
	print "FieldVariable[mpos,ms]:", mops(score), period
	s = time.time()
	testMethodCall()
	period = (time.time() - s) * 1000
	score = period - delta
	print "MethodCall[mpos,ms]:", mops(score), period
	##
	print "misc (it might take a lot of time)"
	s = time.time()
	testFibo(36)
	period = ((time.time() - s) * 1000)
	score = period
	print "fibonacci(36)[mpos,ms]:", mops(score/10), period

run()


